src/poetry/core/version/markers.py

```diff
@@ -284,6 +284,9 @@ def union(self, other: MarkerTypes) -> MarkerTypes:
             if self == other:
                 return self
 
+            if self == other.invert():
+                return AnyMarker()
```

Comment:
Is `AnyMarker` really correct here, or would `EmptyMarker` be more appropriate?

Reply:
That feels backwards to me, but perhaps I am mistaken about how this works.  Isn't an empty marker intended to represent a marker that allows nothing? ie to my understanding the empty marker would be the correct value for the intersection of a marker and its inverse, not to the union of those things.

However I am certainly unsure about the intended relationship between

- `poetry.core.packages.constraints.Constraint`
- `poetry.core.semver.version_constraint.VersionConstraint`
- `poetry.core.semver.version.Version`
- `poetry.core.version.BaseMarker`

some of which seem to have overlapping implementations and purposes.  Not to mention that much of the time the constraint-handling code deals with plain strings and none of the above anyway.  So it's _very_ possible that I'm wrong.  

Perhaps it would be a 'fun' project for someone to either write up the design or tidy this all up...

TLDR: I think what I've done is right but I'm more than willing to be corrected.

Reply:
I think @dimbleby is right: The intersection of a marker and its inverse is empty, but the union of a marker and its inverse is "any" (AnyMarker stands for `"*"`).